### PR TITLE
Add first draft of docs for visitable

### PR DESCRIPTION
**Pull request by lockshaw**
**Description of changes:**

Add high-level documentation of `visitable` with lots of examples and motivation.

**Before merging:**

- [x] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules?

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/890)

### CODE REVIEW

**Comment by lockshaw** [comment]
https://reviewable.io/reviews/flexflow/FlexFlow/890#-

File: lib/utils/README.md [hunk 1]
```diff
@@ -2,19 +2,375 @@
...
+
+  std::string first_name;
+  std::string last_name;
+  int age = 0;
+  bool is_male = false;
+};
+
+Person p{"donald", "knuth", 85};
+```
+would compile without any errors, but is (as of writing this) incorrect.
```
Comment (KateUnger) [inline lib/utils/README.md:64]: I like these examples (the ones for 2.)

File: lib/utils/README.md [hunk 2]
```diff
@@ -2,19 +2,375 @@
...
+      || lhs.last_name != rhs.last_name
+      || lhs.age != rhs.age;
+  }
 
+  std::string first_name;
+  std::string last_name;
+  int age;
+};
 ```
+If we take the previous example of adding an additional `is_male` field to `Person`, it can be easy to miss a location, leading to incorrectness.
```
Comment (KateUnger) [inline lib/utils/README.md:95]: Good examples, but maybe reword to " it can be easy to forget to update the operators."

File: lib/utils/README.md [hunk 3]
```diff
@@ -2,19 +2,375 @@
...
+      && lhs.last_name == rhs.last_name
+      && lhs.age == rhs.age
+      && lhs.is_male == rhs.is_male;
+  }
+
+  friend bool operator!=(Person const &lhs, Person const &rhs) {
+    return lhs.first_name != rhs.first_name 
+      || lhs.last_name != rhs.last_name
+      || lhs.age != rhs.age;
+      // oops, forgot to update here. Have fun debugging :P
```
Comment (KateUnger) [inline lib/utils/README.md:121]: Helpful comment

File: lib/utils/README.md [hunk 4]
```diff
@@ -2,19 +2,375 @@
...
+
+  std::string first_name;
+  std::string last_name;
+  int age;
+  bool is_male;
+};
+```
+and for product types with more fields this grows increasingly tedious to write and maintain. 
+
+4. Hashing: hashing for product types is also relatively trivial, as long as each of the fields is hashable. But again, we have to do a bunch of extra work to specify this, and this work has to be done for each product type in the codebase.
```
Comment (KateUnger) [inline lib/utils/README.md:132]: Reword to "Hashing product types is relatively trivial, long as each of the fields..."

File: lib/utils/README.md [hunk 5]
```diff
@@ -2,19 +2,375 @@
...
+    hash_combine(result, p.first_name);
+    hash_combine(result, p.last_name);
+    hash_combine(result, p.age);
+    hash_combine(result, p.is_male);
+  }
+};
+
+}
+```
+and if we also want to support, say, `std::set<Person>`, we also have to add `operator<`
```
Comment (KateUnger) [inline lib/utils/README.md:181]: I might be missing smt, but what does this operator example have to do with hashing?

File: lib/utils/README.md [hunk 6]
```diff
@@ -2,19 +2,375 @@
...
+FF_VISITABLE_STRUCT(Person, first_name, last_name, age, is_male);
+```
+The key addition here is the calling the `FF_VISITABLE_STRUCT` macro. 
+In addition to defining all of the above functions, this macro also performs a series of `static_assert`s to check that the product type is implemented correctly (for example, it will check that the type is not default constructible).
+The only additional change is the addition of the `req` (which stands for `required`) wrapper on the last field. 
+Conceptually, `req` is simple: it removes default constructibility of the type it wraps (if the last field in the struct is already not default-constructible, no `req` is needed).
+Don't worry if you forget to add a `req`: `FF_VISITABLE_STRUCT` will check that your type properly disables default and partial construction.
+Combined with [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization), we are able to construct a Person as follows:
+```cpp
+Person p = { "donald", "knuth", 85, true };
```
Comment (KateUnger) [inline lib/utils/README.md:263]: Don't feel like this adds much, prob doesnt hurt to leave in tho

File: lib/utils/README.md [hunk 7]
```diff
@@ -2,19 +2,375 @@
...
+template <typename T>
+struct MyLists {
+  std::vector<T> list1;
+  req<std::vector<T>> list2;
+};
+FF_VISITABLE_STRUCT(MyLists, list1, list2); // ERROR
+
+using MyInts = MyLists<int>;
+
+FF_VISITABLE_STRUCT(MyInts, list1, list2); // CORRECT
```
Comment (KateUnger) [inline lib/utils/README.md:285]: An additional ex would help here

File: lib/utils/README.md [hunk 8]
```diff
@@ -2,19 +2,375 @@
...
+    size_t result = 0;
+    hash_combine(result, t.people);
+    hash_combine(result, t.cows);
+    return result;
+  }
+};
+
+}
+```
+which is tedious and bug-prone.
```
Comment (KateUnger) [inline lib/utils/README.md:351]: this is a pretty long example just to say its tedious, that said it gets the point across if you cant think of another way to do so

**Review by wmdi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/890)* status: all files reviewed, 13 unresolved discussions (waiting on @lockshaw and @tarunprabhu)

___
*[`lib/utils/README.md` line 261 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NaXQfvKDMyXl6YpQeCR:-NaXQfvKDMyXl6YpQeCS:undefined) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L261)):*
> ```Markdown
> Conceptually, `req` is simple: it removes default constructibility of the type it wraps (if the last field in the struct is already not default-constructible, no `req` is needed).
> Don't worry if you forget to add a `req`: `FF_VISITABLE_STRUCT` will check that your type properly disables default and partial construction.
> Combined with [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization), we are able to construct a Person as follows:
> ```





_Suggestion:_
```Markdown
Combined with [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization), we are able to construct a `Person` as follows:
```
___
*[`lib/utils/README.md` line 265 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NaXU3xa5IAmnNik_0Jq:-NaXU3xa5IAmnNik_0Jr:bwc1ckr) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L265)):*
> ```Markdown
> Person p = { "donald", "knuth", 85, true };
> ```
> and any subset of the fields would raise an error at compile time.
> ```

Should also mention hashing here.

**Review by tarunprabhu — COMMENTED** [review]
It might be obvious, and it is certainly not relevant to the documentation, but if there are ways to disable default construction of an object in C++, why go down this route? The cases where one needs to have an object be default constructible usually arise with deserialization where one often needs to construct a default object and then initialize the fields with the deserialized values. If we will only ever serialize types over which we have complete control, that same effect can be obtained using a static create method and a protected default constructor. Of course, you have probably thought a lot longer about this, so consider this my <<2 cents.

> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/890#-:-Nah_E4y2jxw6yj2Jz_S:b-4pvcr9)* status: all files reviewed, 21 unresolved discussions (waiting on @KateUnger and @lockshaw)

___
*[`lib/utils/README.md` line 15 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278245280:-NagULG78paf7o8USMY4:bf45n64) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L15)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

"the FlexFlow developers) find undesirable:

1. Default..."

Remove `these are as follows"
</blockquote></details>

Replace parentheses and i.e. with commas. 

... a set of behaviors that we, the FlexFlow developers, find undesirable.
___
*[`lib/utils/README.md` line 28 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278245510:-NahYTXC8-Y0P7JnLOgV:b-eomzz4) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L28)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

I like the explanation below more than the function (to me it's more straight-forward). I would recommend moving the if example below the struct example.
</blockquote></details>

I agree. Just explaining why leaving the fields at their default values may be semantically invalid is sufficient.
___
*[`lib/utils/README.md` line 36 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahZ8KKCDOE0vf2fg3G:-NahZ8KKCDOE0vf2fg3H:b-pz19fw) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L36)):*
> ```Markdown
>   int age = 0;
> }
> ```
> ```

Would it be possible to have an example more relevant to FlexFlow users or developers? Just a thought. This example does well enough to explain why we may want objects to be default constructible. A more relevant example may be more compelling.



_Code quote:_
```Markdown
```cpp
struct Person {
  std::string first_name; // initializes to ""
  std::string last_name; // initializes to ""
  int age = 0;
}
```
```
___
*[`lib/utils/README.md` line 121 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-Nah_7UH2k65ft6a98Y-:-Nah_7UH2k65ft6a98Y0:b-5k1eyz) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L121)):*
> ```Markdown
>       || lhs.last_name != rhs.last_name
>       || lhs.age != rhs.age;
>       // oops, forgot to update here. Have fun debugging :P
> ```

Perhaps mention explicitly that `is_male` was not updated.



_Code quote:_
```Markdown
  // oops, forgot to update here. Have fun debugging :P
```
___
*[`lib/utils/README.md` line 237 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahXZsq8fPh1twQKeCJ:-NahXZsq8fPh1twQKeCK:bsjrnph) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L237)):*
> ```Markdown
> }
> ```
> For a datatype that conceptually is as simple as three independent fields, we now have an rather significant amount of code that must be written and maintained.
> ```

replace with "a significant amount"



_Code quote:_
```Markdown
an rather significant amount
```
___
*[`lib/utils/README.md` line 237 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahXj2F4-JPbZq2fD8z:-NahXj2F4-JPbZq2fD9-:b-ls851s) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L237)):*
> ```Markdown
> }
> ```
> For a datatype that conceptually is as simple as three independent fields, we now have an rather significant amount of code that must be written and maintained.
> ```

Consider rewording. E.g. "Even for such a simple datatype, we have a significant amount of code ..."



_Code quote:_
```Markdown
For a datatype that conceptually is as simple as three independent fields
```
___
*[`lib/utils/README.md` line 257 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahVM1FCY4LbgVtiAOI:-NahVM1FCY4LbgVtiAOJ:b-mz1rfg) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L257)):*
> ```Markdown
> ```
> The key addition here is the calling the `FF_VISITABLE_STRUCT` macro. 
> In addition to defining all of the above functions, this macro also performs a series of `static_assert`s to check that the product type is implemented correctly (for example, it will check that the type is not default constructible).
> ```

Consider replacing with "compile-time checks".
... a series of compile-time checks to ensure that the product type ...



_Code quote:_
```Markdown
`static_assert`s
```
___
*[`lib/utils/README.md` line 257 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahWw7U5ji7oIn3Ww_J:-NahWw7U5ji7oIn3Ww_K:b-shdzg9) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L257)):*
> ```Markdown
> ```
> The key addition here is the calling the `FF_VISITABLE_STRUCT` macro. 
> In addition to defining all of the above functions, this macro also performs a series of `static_assert`s to check that the product type is implemented correctly (for example, it will check that the type is not default constructible).
> ```

Does it do more than just check if the type is not default constructible? If so, would it be worth listing all the properties that the type is expected to have? I am adding review comments from the bottom up, so this probably belongs earlier in the document, but if there is a set of properties that all product types must possess, it should be documented somewhere. If it is, then perhaps there should be a link to the list of properties here so the reader can be clear about what properties the macro checks.



_Code quote:_
```Markdown
is implemented correctly (for example, it will check that the type is not default constructible).
```
___
*[`lib/utils/README.md` line 259 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahVkq58tSWSBBFse9d:-NahVkq64upmkhefev7T:bqtparz) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L259)):*
> ```Markdown
> In addition to defining all of the above functions, this macro also performs a series of `static_assert`s to check that the product type is implemented correctly (for example, it will check that the type is not default constructible).
> The only additional change is the addition of the `req` (which stands for `required`) wrapper on the last field. 
> Conceptually, `req` is simple: it removes default constructibility of the type it wraps (if the last field in the struct is already not default-constructible, no `req` is needed).
> ```

It feels as if `req` is a workaround to force a type to not be default constructable. Does `req` also act as a marker to indicate that some field of a type may not be semantically valid if initialized with a default value?  If so, would it be reasonable advice to require it on all such fields? 



_Code quote:_
```Markdown
Conceptually, `req` is simple: it removes default constructibility of the type it wraps (if the last field in the struct is already not default-constructible, no `req` is needed).
```
___
*[`lib/utils/README.md` line 351 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278254388:-NagZGk2Bs-kx9r8UQbY:b-a2h3mq) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L351)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

this is a pretty long example just to say its tedious, that said it gets the point across if you cant think of another way to do so
</blockquote></details>

I agree with this comment. You could just list the operators that would need to be implemented, and provide a reference to the previous `Person` example where that was done.
___
*[`lib/utils/README.md` line 352 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NagYt_S71i6Szo7TWyA:-NagYt_S71i6Szo7TWyB:b-36psy9) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L352)):*
> ```Markdown
> ```
> which is tedious and bug-prone.
> To remove the constructibility checks performed by `FF_VISITABLE_STRUCT`, we simply use `FF_VISITABLE_STRUCT_NONSTANDARD_CONSTRUCTION` instead:
> ```

You may want to state that this is also necessary for cases where the product type requires a constructor with a specific signature i.e. it is not a copy or move constructor.

**Review by lockshaw — COMMENTED** [review]
I'm not sure I 100% get what you're suggesting, so let me know if the response below leaves something out:

Overall, I'm not against other setups that use static methods, private/protected constructors, etc. The goal of visitable here is really just to make the most desirable behavior (at least in the flexflow context) the easiest to implement. As the example show below, it is possible to get this behavior without using visitable, but it requires 10s-100s of lines of code per struct as opposed to only a couple lines of code, and so in practice the correct behavior will rarely be implemented (even if everyone agrees it _should_ be). In addition, `FF_VISITABLE_STRUCT` bundles not only the correct behaviors, but checks to make sure nothing has been missed, whereas manually implementing things makes it very easy to introduce bugs that fail silently at runtime. 

> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/890#-:-NailluL8zUqKdL6pdTB:b-c3ku02)* status: 0 of 1 files reviewed, 21 unresolved discussions (waiting on @KateUnger, @tarunprabhu, and @wmdi)

___
*[`lib/utils/README.md` line 7 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278245143:-Nc1wCvi6-MJ8_zrp2DM:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L7)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

"referred to as _product types_ in the rest of this document" - delete 'these'
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 15 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278245280:-Nc1wLuZ4qdDNdDRllQ7:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L15)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

Replace parentheses and i.e. with commas. 

... a set of behaviors that we, the FlexFlow developers, find undesirable.
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 28 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278245510:-Nc1xNgu8C2S-KSAv4Rc:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L28)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

I agree. Just explaining why leaving the fields at their default values may be semantically invalid is sufficient.
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 36 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahZ8KKCDOE0vf2fg3G:-NaimwJGB7h3YoBTkrLd:bgcbq2w) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L36)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

Would it be possible to have an example more relevant to FlexFlow users or developers? Just a thought. This example does well enough to explain why we may want objects to be default constructible. A more relevant example may be more compelling.
</blockquote></details>

Maybe, but I wanted to keep the examples as approachable as possible--if people want FlexFlow-relevant examples, they can always just read the FlexFlow code, and adding more complexity just prevents more people from understanding it (and visitable has nothing to do with FlexFlow except for the fact that we use it--it has not fundamental link to machine learning or distributed systems, etc.)
___
*[`lib/utils/README.md` line 64 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278246718:-Nc1xRgHBLr_AnvJYWw9:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L64)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

I like these examples (the ones for 2.)
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 95 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278246960:-Nc1xXe41GhEOLRd2oGy:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L95)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

Good examples, but maybe reword to " it can be easy to forget to update the operators." 
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 121 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-Nah_7UH2k65ft6a98Y-:-Nc1xhYC2qSyuYV42h1u:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L121)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

Perhaps mention explicitly that `is_male` was not updated.
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 121 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278246980:-Nc1xhoD8WSVeaEF8NEq:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L121)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

Helpful comment
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 132 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278247264:-Nc1xnW9-X1mNFUreCRA:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L132)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

Reword to "Hashing product types is relatively trivial, long as each of the fields..."
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 181 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278247466:-Nc1xpZ35SJUdFBVnHu1:b-u0d7m7) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L181)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

I might be missing smt, but what does this operator example have to do with hashing?
</blockquote></details>

`std::unordered_set` uses hashing, and `std:;set` uses ordering (i.e., `operator<`), so if we want to support both types of set we need more than just hashing
___
*[`lib/utils/README.md` line 237 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahXZsq8fPh1twQKeCJ:-Nc1yjz49hbriBMudBy9:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L237)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

replace with "a significant amount"
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 237 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahXj2F4-JPbZq2fD8z:-Nc1yjn_1MlsFoO2nAJC:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L237)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

Consider rewording. E.g. "Even for such a simple datatype, we have a significant amount of code ..."
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 257 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahVM1FCY4LbgVtiAOI:-Nc1yr3IEgChjabf9qMa:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L257)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

Consider replacing with "compile-time checks".
... a series of compile-time checks to ensure that the product type ...
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 257 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahWw7U5ji7oIn3Ww_J:-Nc20gTG4sc_qchcWJTD:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L257)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

Does it do more than just check if the type is not default constructible? If so, would it be worth listing all the properties that the type is expected to have? I am adding review comments from the bottom up, so this probably belongs earlier in the document, but if there is a set of properties that all product types must possess, it should be documented somewhere. If it is, then perhaps there should be a link to the list of properties here so the reader can be clear about what properties the macro checks.
</blockquote></details>

Done.
___
*[`lib/utils/README.md` line 259 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NahVkq58tSWSBBFse9d:-Nain_6F0LqgfjA9nne7:b371bw6) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L259)):*
> It feels as if `req` is a workaround to force a type to not be default constructable.

Yes, this is exactly correct.

> If so, would it be reasonable advice to require it on all such fields?

It's simply not necessary--since only aggregate initialization is allowed in constructing visitable types, as long as the last field is not default constructible the correct behavior is achieved. (also, `req` complicates things such as error messages, and why make error messages worse if you don't have to)
___
*[`lib/utils/README.md` line 261 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NaXQfvKDMyXl6YpQeCR:-Nc20uvlDSvQF2PcSSDQ:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L261)):*
> ```Markdown
> Conceptually, `req` is simple: it removes default constructibility of the type it wraps (if the last field in the struct is already not default-constructible, no `req` is needed).
> Don't worry if you forget to add a `req`: `FF_VISITABLE_STRUCT` will check that your type properly disables default and partial construction.
> Combined with [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization), we are able to construct a Person as follows:
> ```

Done.
___
*[`lib/utils/README.md` line 263 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278249903:-Nc20yAS04Sv6hEhmpYQ:bww2y09) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L263)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

Don't feel like this adds much, prob doesnt hurt to leave in tho
</blockquote></details>

I think it's best to leave in--it's nice to have an example of what to do, rather than just a bunch of examples of what not to do
___
*[`lib/utils/README.md` line 265 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NaXU3xa5IAmnNik_0Jq:-Nc21WR35voEHXF2HyQe:bcu430k) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L265)):*
<details><summary><i>Previously, wmdi (Mengdi Wu) wrote…</i></summary><blockquote>

Should also mention hashing here.
</blockquote></details>

What specifically do you think I should mention here? Just that `Person` is hashable? I've added in some additional text, let me know if you think it's sufficient
___
*[`lib/utils/README.md` line 285 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278250393:-Nc22G3mBZIgeQqrx4MG:b-fznh2y) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L285)):*
<details><summary><i>Previously, KateUnger (Kate Unger) wrote…</i></summary><blockquote>

An additional ex would help here
</blockquote></details>

What aspect of this limitation do you think is insufficiently covered by the existing example? Happy to add more, but I'm not sure what would be good to add
___
*[`lib/utils/README.md` line 351 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278254388:-Nc22_Gj-flMfWVlmNj-:bay6egc) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L351)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

I agree with this comment. You could just list the operators that would need to be implemented, and provide a reference to the previous `Person` example where that was done.
</blockquote></details>

Added in a disclaimer so people can just skim the code if they want
___
*[`lib/utils/README.md` line 352 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#-NagYt_S71i6Szo7TWyA:-Nc23XhW2j17t2Pjb0MG:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L352)):*
<details><summary><i>Previously, tarunprabhu (Tarun Prabhu) wrote…</i></summary><blockquote>

You may want to state that this is also necessary for cases where the product type requires a constructor with a specific signature i.e. it is not a copy or move constructor.
</blockquote></details>

Done.

**Review by tarunprabhu — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/890)* status: all files reviewed, 11 unresolved discussions (waiting on @KateUnger)

**Review by wmdi — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/890)* status: all files reviewed, 11 unresolved discussions (waiting on @KateUnger)

**Review by KateUnger — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/890)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @lockshaw)

___
*[`lib/utils/README.md` line 285 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/890#gh-1278250393:-NcEnfNh-rylMO3V9nle:b-jpv98u) ([raw file](https://github.com/flexflow/FlexFlow/blob/85ca60bef191e36ef7975e7d42217f5f6108dc91/lib/utils/README.md#L285)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

What aspect of this limitation do you think is insufficiently covered by the existing example? Happy to add more, but I'm not sure what would be good to add
</blockquote></details>

Rereading it, I think you're good
